addons/mail/models/mail_thread.py

```diff
@@ -925,143 +910,92 @@ def message_route(self, message, message_dict, model=None, thread_id=None, custo
             raise TypeError('message must be an email.message.Message at this point')
         MailMessage = self.env['mail.message']
         Alias = self.env['mail.alias']
-        fallback_model = model
 
-        # Get email.message.Message variables for future processing
+        # various information
+        fallback_model = model
+        local_hostname = socket.gethostname()
         message_id = message.get('Message-Id')
-        email_from = decode_header(message, 'From')
-        email_to = decode_header(message, 'To')
-        references = decode_header(message, 'References')
-        in_reply_to = decode_header(message, 'In-Reply-To').strip()
+        # compute references to find if message is a reply to an existing thread
+        references = tools.decode_smtp_headers(message, 'References')
+        in_reply_to = tools.decode_smtp_headers(message, 'In-Reply-To').strip()
         thread_references = references or in_reply_to
+        reply_match, reply_model, reply_thread_id, reply_hostname = tools.email_references(thread_references)
+        # author and recipients
+        email_from = tools.decode_smtp_headers(message, 'From')
```

Comment:
Some headers seem to be decoded often, wouldn't it be possible to eagerly decode (some) SMTP headers?